src/java.base/share/classes/java/io/FileInputStream.java

```diff
@@ -236,14 +237,39 @@ public int read() throws IOException {
 
     private native int read0() throws IOException;
 
+    private native int readBytes0(byte[] b, int off, int len) throws IOException;
+
     /**
      * Reads a subarray as a sequence of bytes.
      * @param     b the data to be written
      * @param     off the start offset in the data
      * @param     len the number of bytes that are written
      * @throws    IOException If an I/O error has occurred.
      */
-    private native int readBytes(byte[] b, int off, int len) throws IOException;
+    private int readBytes(byte[] b, int off, int len) throws IOException {
+        Objects.checkFromIndexSize(off, len, b.length);
+        final int start = off;
+        long comp = Blocker.begin();
+        try {
+            while (len > 0) {
+                int n = readBytes0(b, off, len);
+                if (n < 0) {
+                    if (off == start)
+                        return -1;
+                    break;
+                }
+                off += n;
+                len -= n;
+            }
+        } catch (IOException e) {
+            // Throw only if no bytes have been read
+            if (off == start)
```

Comment:
If i am reading code correctly then with the new implementation, until client issue the next "FIS.read" he may or may not know if there was exception pending in previous 'read' call ?.
I am not sure in case of partial read we have suppress the exception.

Reply:
> If i am reading code correctly then with the new implementation, until client issue the next "FIS.read" he may or may not know if there was exception pending in previous 'read' call ?.
> I am not sure in case of partial read we have suppress the exception.

It's a forced move because bytes have already been copied into user's byte[].